daemon/src/main/scala/io/gearpump/cluster/worker/Worker.scala

```diff
@@ -437,18 +439,23 @@ private[cluster] object Worker {
       val format = new java.text.SimpleDateFormat(datePattern)
       format.format(timestamp)
     }
-  }
 
-  private def filterDaemonLib(classPath: Array[String]): Array[String] = {
-    val jarFile = new File(getClass.getProtectionDomain.getCodeSource.getLocation.toURI.getPath)
-    val daemonDir = jarFile.getParent
-    classPath.filterNot{path =>
-      //The class path maybe a symbol link
-      val canonicalized = Try(new File(path).getCanonicalPath)
-      if(canonicalized.isSuccess && canonicalized.get.startsWith(daemonDir)) {
-        true
-      } else {
-        false
+    private def filterDaemonLib(classPath: Array[String]): Array[String] = {
+      val classpathWhiteList = getClasspathWhiteList(config)
+
+      def inWhiteList(file: File): Boolean = {
+        //Resolve the file path's symbol link, the function may throw exception when resolving failed
+        val canonicalPath = Try(file.getCanonicalPath)
+        canonicalPath.isSuccess && classpathWhiteList.contains(canonicalPath.get)
+      }
+
+      classPath.filter{ path =>
+        //The path may represent a file or folder
+        if(inWhiteList(new File(path)) || inWhiteList(new File(path).getParentFile)) {
```

Comment:
The expression itself indicates a boolean value, no need to return true/false explicitly.